examples/05_tool_inference_lock/ex_tool_inference_lock/inference_lock_helper.rb

```diff
@@ -0,0 +1,130 @@
+module Examples
+  module ToolInferenceLock
+    # Helper object to lock tool inference using Shift or arrow keys.
+    class InferenceLockHelper
+      # Create InferenceLockHelper object.
+      def initialize
+        @axis_lock = nil
+        @mouse_x = nil
+        @mouse_y = nil
+      end
+
+      # Call this method from +onKeyDown+.
+      #
+      # @param key [Integer]
+      # @param view [Sketchup::View]
+      # @param tool [Object]
+      # @param active_ip [Sketchup::InputPoint]
+      #   The InputPoint currently used to pick points.
+      # @param reference_ip [Sketchup::InputPoint]
+      #   An InputPoint marking where the current operation was started.
+      def handle_keydown(key, view, tool, active_ip, reference_ip = active_ip)
+        if key == CONSTRAIN_MODIFIER_KEY
+          try_lock_constraint(view, active_ip)
+        else
+          try_lock_axis(key, view, reference_ip)
+        end
+
+        # Emulate mouse move to let tool do its InputPoint pick and perform any
+        # other actions the tool may do along with it.
+        tool.onMouseMove(0, @mouse_x, @mouse_y, view) if @mouse_x && @mouse_y
+        view.invalidate
+      end
+
+      # Call this method from +onKeyUp+.
+      #
+      # @param key [Integer]
+      # @param view [Sketchup::View]
+      # @param tool [Object]
+      def handle_keyup(key, view, tool)
+        return unless key == CONSTRAIN_MODIFIER_KEY
+        return if @axis_lock
+
+        # Calling this method with no argument unlocks inference.
+        view.lock_inference
+        
+        # Emulate mouse move to update InputPoint picking.
+        tool.onMouseMove(0, @mouse_x, @mouse_y, view) if @mouse_x && @mouse_y
```

Comment:
I'm still troubled by this coupling. I'd rather see an event used, where the tool can register what should happen. This helper shouldn't be meddling with calls directly back to it's owning tool.

Reply:
I've changed it so the tool itself caches the current mouse position and has broken out the input point picking code to a method called both on mouse move and key events.

Reply:
This looks much better!